uploadservice/src/main/java/net/gotev/uploadservice/extensions/ContextExtensions.kt

```diff
@@ -30,7 +33,60 @@ fun Context.startNewUpload(
         putExtra(taskNotificationConfig, notificationConfig)
     }
 
-    startService(intent)
+    try {
+        /*
+        When trying to start a service on API 26+
+        while the app is in the background, an IllegalStateException will be fired
+
+        https://developer.android.com/reference/android/content/Context#startService(android.content.Intent)
+
+        Then why not using startForegroundService always on API 26+? Read below
+         */
+        startService(intent)
+    } catch (exc: Throwable) {
+        if (Build.VERSION.SDK_INT >= 26 && exc is IllegalStateException) {
+            /*
+            this is a bugged Android API and Google is not going to fix it
+
+            https://issuetracker.google.com/issues/76112072
+
+            Android SDK can not guarantee that the service is going to be started in under 5 seconds
+            which in turn can cause the non catchable
+
+            RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground()
+
+            so the library is going to use this bugged API only as a last resort, to be able
+            to support starting uploads also when the app is in the background, but preventing
+            non catchable exceptions when you launch uploads while the app is in foreground.
+
+            As a supplementary safety measure, the library is going to try to catch the
+            RemoteServiceException when using the unsafe API,
+            logging a BuggedAndroidServiceAPIException which you can also handle in a custom way
+            by setting a custom logger delegate with:
+
+            UploadServiceLogger.setDelegate
+             */
+
+            val defaultExceptionHandler = Thread.getDefaultUncaughtExceptionHandler()
+
+            if (defaultExceptionHandler !is RemoteServiceExceptionHandler) {
+                Thread.setDefaultUncaughtExceptionHandler(
+                    RemoteServiceExceptionHandler(defaultExceptionHandler)
+                )
+            }
+
+            startForegroundService(intent)
```

Comment:
This is confusing to me because it looks like you are trying to catch the error before it can actually happen. Calling `startForegroundService` on line 78 here is what would cause the `RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground()` error, right?

I don't have much experience with Java so I could be reading it wrong.

Reply:
Some reference: https://developer.android.com/reference/java/lang/Thread.UncaughtExceptionHandler